src/coreclr/tools/Common/TypeSystem/Common/CastingHelper.TypeEquivalence.cs

```diff
@@ -0,0 +1,312 @@
+                }
+                else if (type1.IsValueType)
+                {
+                    return CompareStructuresForEquivalence(type1, type2, visited, enumMode: false);
+                }
+                else if ((type1.IsDelegate == type2.IsDelegate) && type1.IsDelegate)
+                {
+                    return CompareDelegatesForEquivalence(type1, type2, visited);
+                }
+
+                return false;
+            }
+
+            static bool CompareDelegatesForEquivalence(MetadataType type1, MetadataType type2, StackOverflowProtect visited)
+            {
+                var invoke1 = type1.GetMethod("Invoke", null);
+                var invoke2 = type2.GetMethod("Invoke", null);
+
+                if (invoke1 == null)
+                    return false;
+
+                if (invoke2 == null)
+                    return false;
+
+                return invoke1.Signature.EquivalentTo(invoke2.Signature, visited);
+            }
+
+            static bool CompareStructuresForEquivalence(MetadataType type1, MetadataType type2, StackOverflowProtect visited, bool enumMode)
+            {
+                foreach (var method in type1.GetMethods())
+                {
+                    // If there are any methods, then it isn't actually a type-equivalent type
+                    return false;
+                }
+
+                foreach (var method in type2.GetMethods())
+                {
+                    // If there are any methods, then it isn't actually a type-equivalent type
+                    return false;
+                }
+
+                // Compare field types for equivalence
+                var fields1 = type1.GetFields().GetEnumerator();
+                var fields2 = type2.GetFields().GetEnumerator();
+
+                while (true)
+                {
+                    bool nonTypeEquivalentValidFieldFound;
+
+                    FieldDesc field1 = GetNextTypeEquivalentField(fields1, enumMode, out nonTypeEquivalentValidFieldFound);
+                    if (nonTypeEquivalentValidFieldFound)
+                        return false;
+                    FieldDesc field2 = GetNextTypeEquivalentField(fields2, enumMode, out nonTypeEquivalentValidFieldFound);
+                    if (nonTypeEquivalentValidFieldFound)
+                        return false;
+
+                    if ((field1 == null) && (field2 == null))
+                    {
+                        // We ran out of fields on both types before finding a failure
+                        break;
+                    }
+
+                    if ((field1 == null) || (field2 == null))
+                    {
+                        // we ran out of fields on 1 type.
+                        return false;
+                    }
+
+                    // Compare the field signatures for equivalence
+                    // TODO: Technically this comparison should include custom modifiers on the field signatures
```

Comment:
Is this a real possibility that we should track with an issue?

Reply:
Eh, no. We don't even check field types on fields correctly today for normal field references, much less custom modifiers.